### ISSUE TITLE
Rechunk errors if given 0 as a chunk size

### ISSUE THREAD

**Issue by drwelby** [dask/dask#3501]
Zeros are valid as placeholders in defining a chunk size when creating an array, but throw an error if used similarly within a rechunk size. Zero-width chunks can also be generated by a mix of slicing and concatenating.

```python
import dask, numpy
a = numpy.ones([100,100])
b = dask.array.from_array(a, chunks=((50,50),(100,0))) #ok
c = b.rechunk(((40,60), (100,0))) #AssertionError
```
The error is thrown on this line:

```python
assert new_idx == tuple(len(c) - 1 for c in chunks)
````

**Comment by mrocklin** [comment]
I agree that this inconsistency is probably undesirable.  Thank you @drwelby for such a clear issue and failing test case.

I'm inclined to allow 0-length chunks in rechunk, but we should probably let this settle for a moment to see if others have thoughts or concerns.

@drwelby is this something that you would be willing to fix if people are ok with this change?

**Comment by drwelby** [comment]
I believe I could fix this. It's an odd edge case so it can sit to see what feedback it generates.

**Comment by shoyer** [comment]
On the contrary, I think we should try to maintain the invariant that `array.rechunk(chunks).chunks == chunks`. If that means making empty chunks, then so be it. Potentially we could add another method for cleaning up/removing zero-width chunks.

**Comment by drwelby** [comment]
@shoyer Good point. I tried just commenting out the assert statement that's creating the error and all the `dask.array` tests pass. I also tried my example above and the invariant test and they work too.

So should we just drop the assertion or try to make it work with zero-width chunks?

**Comment by mrocklin** [comment]
Yes, it sounds like dropping the assertion seems sensible.  Are you up for submitting a PR with this change @drwelby ?

(also, sorry for the long delay in response)
